### ISSUE TITLE
adapter: improve launchdarkly sync dataflow

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#17964]
LaunchDarkly interacts weirdly with `ALTER SYSTEM SET` and the stash. Some of the current workflow is due to each of these 3 things being developed kinda separately. We need to consolidate them and have a better story around what exactly `ALTER SYSTEM SET` is for and when/how it should override LD, and what to do in that case. Some slack threads copied for context:

----

> Other ideas are: move the launchdarkly calls around so it's available as a data source before init'ing the stash.

I think this is the cleanest way to do this at the moment.
You would need to evolve SystemParameterFrontend as follows:
- Add a method that attempts to establish a connection, but stops after some timeout.
- Add a method that fetches tries to fetch a value of a specific SystemVar from a given &str and returns Option<T> if the connection is already established.

You can then attempt early initialization and value lookup and fallback to the default if LD does not respond for some reason.

Regarding all the other questions in the discussion after that, I'll try to summarize below.
- The current design is best illustrated [on this picture](https://www.notion.so/materialize/Dynamic-environment-configuration-ad11421b40934e4cba675bbe550cc2b2?pvs=4#0492ff9297f34fc1a10196246865305f). LaunchDarkly has been added as a frontend to our existing mz_system-based ALTER SYSTEM API. LD maintains a source of truth and stash acts as a persistent caching layer so we can continue operating upon restarts even if LD is down.
- We have designed the workflow carefully in order to minimize the places where we have a hard dependency on LD.
- At the moment, a connection to LD is only required once when creating the environmentd instance ([code](https://github.com/MaterializeInc/materialize/blob/51cad9394089abf62b716f4babf2dfbedf74480b/src/adapter/src/catalog.rs#L2885-L2960)). The enclosing load_system_configuration method will tell you how SystemVars are populated and which source wins. In subsequent restarts, this whole blocked will be skipped because self.state.system_config().config_has_synced_once() will return true.
- In general, LD sync works as unidirectional dataflow loop where state changes from LD (which acts as a source of truth) are propagated and persisted to stash every 15 seconds ([code](https://github.com/MaterializeInc/materialize/blob/51cad9394089abf62b716f4babf2dfbedf74480b/src/adapter/src/config/mod.rs#L53-L59) / [design doc section](https://www.notion.so/materialize/Dynamic-environment-configuration-ad11421b40934e4cba675bbe550cc2b2?pvs=4#235711ef3c7e4b82b08efce59c6fc001)). This means that if you run ALTER SYSTEM SET locally between loops, that value is going to be overridden when the loop ticks again.

IIUC the main use case that we have at the moment is turning RBAC on and off as a self-service. I think the command handler for this action should actually contact LD using its REST API and change the targeting rule for the environments under the organization_id that issues the request. If LD is not offline, that change will be available in all environments after at most 15 seconds. If LD is down for some reason, the command handler should report an error and then support can run ALTER SYSTEM SET manually on the specified machines.

> Maybe LD should always read the values from the stash instead of storing them locally.

I think this violates the current design assumption that LD is the source of truth and stash acts is a persistent environmentd-specific cache. Changes to the LD source of truth are not necessarily always a result of an ALTER SYSTEM SET call. For example, a PM can enable some feature toggle for a population of environments that includes your environment. Another option is to not list include the rbac_enabled feature flag under SystemVars::iter_synced. Then it won't be synchronized from LD and every user will be able to toggle this feature directly using ALTER SYSTEM SET.

> I agree though that the current situation seems wrong where the effective value of a parameter can change after a restart.

I agree that this is somewhat confusing, but I don't thing it is wrong. We need this option for the unlikely situation where LD goes down. We can do better here by rejecting sequence_alter_system_set calls for parameters that are managed by LD if a connection is alive.

> It seems like ALTER SYSTEM should also update LD, if it doesn't already.

This is another way to do it, but we need to be careful to do this outside of SessionClient::set_system_vars because this API is shared with the LD sync loop (see SystemParameterBackend::push). Also, unless we buy a LD seat for each organization_id, these changes will be logged under one generic user.

**Comment by mjibson** [comment]
My thoughts: I'm not convinced we want the behavior of the system to change based on if LD is up or not. Having envd change its behavior after restart due to differences between LD and `ALTER SYSTEM SET` also seems quite wrong, even if we have to accept an uptime dependency on LD.

**Comment by aalexandrov** [comment]
cc @benesch it seems that the current design of piggy-backing on `ALTER SYSTEM` for the backend when running the LD sync loop is causing some confusion.

**Comment by benesch** [comment]
Sorry, having trouble following. What is the actual problem with the current design?

**Comment by aalexandrov** [comment]
I think that there is some confusion that `ALTER SYSTEM` can be accessed directly through the internal SQL port, but for most of the entries changes issued that way will be short-lived because the next tick in the LD sync loop will override that value.

There might be other concerns, but I'll let @mjibson and @jkosh44 list them.

**Comment by mjibson** [comment]
The other meta concern is that no one on the adapter team knows how this works, and we were messing with adjacent stuff and got tripped up over our lack of understanding.

**Comment by jkosh44** [comment]
Ultimately I think the current behavior is probably fine as long as we're aware of it. However, I have two concerns/thoughts:

---

The following scenario is currently possible, at best it's weird and at worst it's wrong from a consistency point of view:
  1. Someone sets the value of system variable `var` to `v_0` in LD.  
  2. Someone alters the value of system variable `var` to `v_1` via `ALTER SYSTEM SET`.
  3. Everyone observes the value to be `v_1` and the system behaves as if the value is `v_1`.
  4. No one does anything for 15 minutes and LD refreshes the variable.
  5. Everyone observes the value to be `v_0` and the system behaves as if the value is `v_0`.

Specifically step 5 is at least confusing because the value of `var` changed even though no one changed it. Though it may also be incorrect from a consistency point of view.

---

> > Maybe LD should always read the values from the stash instead of storing them locally.
>
> I think this violates the current design assumption that LD is the source of truth and stash acts is a persistent environmentd-specific cache.

Without being familiar with the LD API, this sounds slightly backwards to me. I would have thought that the stash is the source of truth and LD could act as a cache for people using the LD frontend. If environmentd and the stash are available but LD is not available, then you can alter the values of configuration variables AND view the affects of that change. If environmentd and the stash are not available but LD is available, then you can alter the values of configuration variables but never actually view the affects of that change until environmentd and the stash become available again.

In general I get a little nervous anytime the source of truth for the state of Materialize is spread out to multiple systems.

**Comment by aalexandrov** [comment]
> No one does anything for 15 minutes and LD refreshes the variable.
nit: the sync period is 15 seconds not 15 minutes, and if this helps I think we can make it 1 second without stressing the network or the CPU of the running instances too much.

**Comment by benesch** [comment]
Chatted with @jkosh44 about this today before I managed to type all the below out.

---

> Without being familiar with the LD API, this sounds slightly backwards to me. I would have thought that the stash is the source of truth and LD could act as a cache for people using the LD frontend. If environmentd and the stash are available but LD is not available, then you can alter the values of configuration variables AND view the affects of that change. If environmentd and the stash are not available but LD is available, then you can alter the values of configuration variables but never actually view the affects of that change until environmentd and the stash become available again.
> 
> In general I get a little nervous anytime the source of truth for the state of Materialize is spread out to multiple systems.

LD doesn't work like that, I'm afraid. The LD UI always uses the LD service as the source of truth. You can't have the LD UI read out of the Materialize stashes, for example.

----

Here's how I think about it:

  * All configuration parameters in Materialize are in a global namespace and managed variously by `SHOW`, `SET`, and `ALTER SYSTEM`.
  * Some parameters are "internal". They are only visible via `SHOW` when logged in to an internal port. They can only be altered via `ALTER SYSTEM` when logged in to an internal port.
  * Some parameters are automatically synced with LD. These parameters are _always_ internal parameters. We'd never use LD to sync a parameter that is settable by end users. These parameters are settable via `ALTER SYSTEM` on the internal port in case of emergency.
  * Non internal parameters may be set with `ALTER SYSTEM`, to apply a new default for all sessions, or with `SET`, to apply a value for the current session only.

I think what would help in resolving this confusion is to come up with the set of properties that describe each parameter. Here's one shot at that:

```
/// Determines who can update a parameter.
enum Writability {
    /// No one can update this parameter. Its value is fixed at compile time.
    None,
    /// The parameter is synced with an external system (e.g., LaunchDarkly). It can be set
    /// via `ALTER SYSTEM` on an internal port in an emergency.
    Synced,
    /// The parameter can only be set via `ALTER SYSTEM` via an internal port.
    Internal,
    /// The parameter can be set via `ALTER SYSTEM` via an internal port,
    /// `ALTER SYSTEM` via an external port if the user is a superuser, or
    /// `SET` on a per-session basis.
    All,
}

/// Determines who can view a parameter's value.
enum Readability {
    /// The parameter is only visible via the internal ports.
    Internal,
    /// The parameter is visible via all ports.
    All,
}
```

**Comment by aalexandrov** [comment]
> Some parameters are automatically synced with LD. These parameters are always internal parameters. 

@benesch this is not the case at the moment. For example all limit parameters are visible via `SHOW`.

**Comment by jkosh44** [comment]
I like the proposal and think that we should modify the code the more explicitly match that structure. Though I don't think it's a high priority right now.

> ```Rust
> /// Determines who can update a parameter.
> enum Writability {
>    /// No one can update this parameter. Its value is fixed at compile time.
>    None,
>    /// The parameter is synced with an external system (e.g., LaunchDarkly). It can be set
>    /// via `ALTER SYSTEM` on an internal port in an emergency.
>    Synced,
>    /// The parameter can only be set via `ALTER SYSTEM` via an internal port.
>    Internal,
>    /// The parameter can be set via `ALTER SYSTEM` via an internal port,
>    /// `ALTER SYSTEM` via an external port if the user is a superuser, or
>    /// `SET` on a per-session basis.
>    All,
>}
```

Do you have an examle of a `Writability::Internal` variable? I can't think of a use case for a variable that is only write-able via the internal port but not managed by LD.

**Comment by benesch** [comment]
> Note that the current code does not enforce this constraint for "internal" parameters (for example limit parameters are both synced and visible through the public port):

Yep, I got that wrong in prose, but I think the `Readability` and `Writability` enums allow the limit parameters to be properly represented!

**Comment by aalexandrov** [comment]
> Do you have an example of a `Writability::Internal` variable?

I think @guswynn has one in #18010. Another one is `has_been_synced_once`, which is an internal flag used to determine whether `Catalog::open` needs to block until the LD state has been synced at least once with stash.

**Comment by jkosh44** [comment]
> > Do you have an example of a `Writability::Internal` variable?
> 
> I think @guswynn has one in #18010. Another one is `config_has_synced_once`, which is an internal flag used to determine whether `Catalog::open` needs to block until the LD state has been synced at least once with stash.
> 
> Basically `Writability::Internal` is now represented as `SystemVars::iter() - SystemVars::iter_synced()`, while `Writability::Synced` is now `SystemVars::iter_synced()`.

Gus wrote:
> adds the system parameter (which will later be connected to a LaunchDarkly flag),

which makes it sound it's more like `Writability::Synced` since it's connected to LD. However, I did think of an existing use-case, testing variables. Like `mock_audit_event_timestamp`.

**Comment by mjibson** [comment]
> These parameters are settable via ALTER SYSTEM on the internal port in case of emergency.

This doesn't say what will happen at the 15s LD sync, which will override this even in case of emergency.

If the emergency was the LD is down so we can't change things in LD's UI, then as soon as LD comes back up the 15s loop will overwrite what we emergency set until we can get into the LD UI and change it back, which would be 3 var changes the user experiences (1: our emergency ALTER SYSTEM, 2: LD coming back online with the old value, 3: us changing LD to the new value and waiting for it to sync).

If the emergency is something else, then executing ALTER SYSTEM will get reverted in 15s.

**Comment by aalexandrov** [comment]
> This doesn't say what will happen at the 15s LD sync, which will override this even in case of emergency.

@mjibson You are correct. The handling of a "LD down" scenario is somewhat convoluted. For unblocking new environments, the proposed workflow is in this comment:

https://github.com/MaterializeInc/materialize/blob/e973a6c01aec892dc5d6c441b61a8664b2bd6f1c/src/adapter/src/catalog.rs#L2914-L2927

For existing environments, as you mention:

1. You can override a system var through an `ALTER SYSTSEM SET` from the internal port.
2. You can update the parameter from the LD frontend. This will bring the LD config in sync with the environmentd which is changed manually in (1).
3. As soon as the LD client re-establishes the connection, your value will be overridden from the LD config. If for some reason (2) was not possible, you will have the value original value, which might be quite surprising.

I think there are several solutions to improve this workflow. For example:

**Option 1**: One can have a dedicated internal `SystemVar` called `CONFIG_SYNCED_ENABLED` to guard the config sync loop

https://github.com/MaterializeInc/materialize/blob/4bd170b16c3b4b42eccb4df854c1928bf85f8cb4/src/adapter/src/config/mod.rs#L53-L59

**Option 2**: One can have an admin command that installs some Cilium rules that block connectivity to LD. That way Materialize ops can control the exact point when the connection is re-enabled and perform step (2) from the above workflow before that.

**Comment by aalexandrov** [comment]
To summarize: 

1. The current workflow comes with some simplicity that was modeled after the happy case. 
2. This comes with the price of some complication that affect some workflows when LD is down (starting a new `environmentd`, changing the values of synced parameter values).
3. So far we have not been affected by incident where LD is down. You can look at the [uptime history](https://status.launchdarkly.com/uptime) to gauge how many of those we should expect, but if last year is a good predictor, I think we'll have to account for ~1 incident every 3 months.
4. Based the expected frequency of incidents (3) I think it makes more sense to think about mechanisms to further optimize the workflows in (2) without complicating the workflows in (1).

**Comment by mjibson** [comment]
A more specific proposal:

LD sync indeed now does not use `ALTER SYSTEM` or any other SQL, rather it makes some specific coordinator messages that describe the same thing, and those get converted directly into catalog operations for a catalog txn. Thus, the only possible user of `ALTER SYSTEM` is a support engineer.

I propose we add a new layer to the vars. It is set via `ALTER SYSTEM SET|RESET`. It is persisted to disk in a separate location in the stash from the existing LD vars. It takes precedent over any LD vars.

This will allow support engineers to do anything they need in case of emergency, and has deterministic behavior that is not dependent on the LD sync timing or restarts.

**Comment by benesch** [comment]
What's the problem that the proposal is solving? IMO, the fact that LD overrides any changes made manually via `ALTER SYSTEM` is a feature, not a bug. I'd like to be able to trust that, barring sync errors, what's shown in the LD UI is in fact the value of the flag that is active in the environment. If we do the proposal, we'll lose that property. Someone is going to leave an override sitting in a customer environment, and someone else will wind up very confused about why their LD config parameter is not applying. It won't be obvious in the LD UI at all that the flag has been overridden.

I expect that we'll ~never need to use the `ALTER SYSTEM` workaround when LD is down. If LD goes down, for ~all feature flags (and new environment spinup), we can just wait. I think it will be incredibly rare that we have an urgent need to twiddle an LD flag, and the probability of that coinciding with LD downtime is very small.

**Comment by aalexandrov** [comment]
> I expect that we'll ~never need to use the ALTER SYSTEM workaround when LD is down. If LD goes down, for ~all feature flags (and new environment spinup), we can just wait.

This is probably true right now, but it might become a problem in the future, so there is some value in thinking through some solutions.

> If we do the proposal, we'll lose that property. Someone is going to leave an override sitting in a customer environment, and someone else will wind up very confused about why their LD config parameter is not applying. It won't be obvious in the LD UI at all that the flag has been overridden.

I think the point of @mjibson proposal is that such scenarios can handled more easily by an `mzadmin` command that provides an overview of settings written in that layer and the ability to (bulk) reset them. The current workflow would be:

1. Each environment needs to be restarted or the current `config` code tweaked such that in LD sync does not auto-resume for "dirty" environments.
2. On-call and support engineers will have to keep track of all "dirty" values per environment.
3. Resuming normal operations will require a manual process that first sets all values from (2) in LD, resets them locally, and only after that reverts (1).

**Comment by mjibson** [comment]
> What's the problem that the proposal is solving?

Problems:

1. If LD is down `ALTER SYSTEM` will get reverted to its LD setting as soon as LD comes back up. Presumably we needed to run `ALTER SYSTEM` for a user need, so as soon as LD is back up their problem comes back during the sync loop. This was described in my "three changes" comment at https://github.com/MaterializeInc/materialize/issues/17964#issuecomment-1467521909.
2. If LD is not down and `ALTER SYSTEM` is executed (which could absolutely be done by one of us), it will have unexpected behavior (returns success, but then overwritten in <15s, so the person who ran it gets very confused).

In both cases `ALTER SYSTEM` doesn't have the expected behavior because it gets overwritten as soon as LD comes back online or does a regular sync.

**Comment by mjibson** [comment]
I am fine changing nothing and instead making sure our support team is aware of:

> as soon as LD comes back up the 15s loop will overwrite what we emergency set until we can get into the LD UI and change it back, which would be 3 var changes the user experiences (1: our emergency ALTER SYSTEM, 2: LD coming back online with the old value, 3: us changing LD to the new value and waiting for it to sync).

as the expected workflow for changing a system var during LD downtime.

**Comment by aalexandrov** [comment]
@materializesupport / @ggnall note [the above comment :arrow_up:](https://github.com/MaterializeInc/materialize/issues/17964#issuecomment-1470478445) and let us know where best to make this information more prominently visible so support members are not surprised by this quirk.

**Comment by benesch** [comment]
I'd say that support should ~never be running `ALTER SYSTEM` directly! I'd limit this to an emergency playbook for the on-call engineer. I'm not sure we have such a playbook, though.
